### PR TITLE
feat(npm): support custom registryUrls

### LINKED ISSUE #10110: Remove npmrc from datasource arguments
**What would you like Renovate to be able to do?**

Remove `npmrc` as a field from datasource lookups.

**Did you already have any implementation ideas?**

Any custom `registryUrls` need to be included as part of config, during the extract phase. Auth should be in `hostRules`.

### PR DESCRIPTION
**Pull request by rarkins**
## Changes

Refactors `npmrc` handling from `datasource/npm` into `datasource/index` so that custom `registryUrls` can be supported.

## Context

Closes #10110

Helps #9145

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by rarkins** [comment]
@viceice this was done in 5 separate commits, each of which passed tests

File: lib/modules/datasource/npm/get.spec.ts [hunk 1]
```diff
@@ -146,7 +150,8 @@ describe('modules/datasource/npm/get', () => {
       .get('/renovate')
       .reply(200, { name: 'renovate' });
     setNpmrc(npmrc);
-    await getDependency(http, 'renovate');
+    const registryUrl = resolveRegistryUrl('renovate');
+    await getDependency(http, registryUrl, 'renovate');
```
Comment (viceice) [inline lib/modules/datasource/npm/get.spec.ts:154]: Shouldn't we add an expectation for the result?
The line below will be removed soon

**Review by viceice — COMMENTED** [review]
Otherwise LGTM

File: lib/modules/datasource/index.ts [hunk 2]
```diff
@@ -239,15 +241,24 @@ async function fetchReleases(
...
     return null;
   }
+  if (datasourceName === 'npm') {
+    if (is.string(config.npmrc)) {
+      setNpmrc(config.npmrc);
+    }
+    if (!is.nonEmptyArray(registryUrls)) {
+      registryUrls = [resolveRegistryUrl(config.packageName)];
+    }
+  }
```
[the remark this answers is not in the document]
Comment (rarkins) [reply lib/modules/datasource/index.ts:256]: Yes, eventually we will push it all back to the manager/extract

File: lib/modules/datasource/npm/get.spec.ts (hunk 1 above)
Comment (rarkins) [reply lib/modules/datasource/npm/get.spec.ts:154]: I guess so, but I was just keeping up the status quo

**Review by viceice — APPROVED** [review]

**Pull request merged**